File: internal/cryptoid/kek.go
package cryptoid

// MASTER_KEK → per-project KEK → DEK wrap (SPEC §2.2–§2.5) — Go tarafı.
// Server-decrypt pivotunda zarf kriptosu SUNUCUDA koşar (§2.7); bu dosya
// yalnızca DR aracının (§8.4: `wapps dr restore`) ihtiyacı olan alt kümeyi
// taşır: kid türetimi, HKDF per-project KEK, WKW1 DEK-unwrap. Worker'daki
// crypto/kek.ts ile bayt-bayt paritelidir; frozen vektörler her iki tarafta
// CI'da koşar (build gate 2) — sapma release-blocker'dır.

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"io"

	"golang.org/x/crypto/chacha20poly1305"
	"golang.org/x/crypto/hkdf"
)

// WrapRecipient, v2 manifest'lerdeki tek geçerli wrap alıcısı (kapalı küme, §2.4).
const WrapRecipient = "worker-kek:v1"

// WrapMagic, WKW1 çerçeve sihirli baytları (§2.4).
const WrapMagic = "WKW1"

// hkdfSalt, per-project KEK türetiminin sabit salt'ı (20 ASCII bayt, §2.3).
const hkdfSalt = "wapps-secrets/kek/v1"

// WrapTotalLen, bir WKW1 wrap'inin toplam bayt uzunluğu: magic(4) + nonce(24)
// + ciphertext(32+16) = 76 (§2.4).
const WrapTotalLen = 4 + chacha20poly1305.NonceSizeX + dekSize + chacha20poly1305.Overhead

// ErrWrapInvalid, WKW1 çerçeve/AEAD ihlali — tamper veya anahtar uyuşmazlığı
// (§2.4 WRAP_INVALID, fail-closed).
var ErrWrapInvalid = fmt.Errorf("cryptoid: WRAP_INVALID")

// KekKid, bir master anahtarın kid'ini türetir: SHA-256(32 HAM bayt)'ın ilk
// 16 küçük-harf hex karakteri (§2.2 — girdi ASLA ASCII hex string'i değildir).
func KekKid(raw []byte) (string, error) {
	if len(raw) != 32 {
		return "", fmt.Errorf("cryptoid.KekKid: master key must be 32 raw bytes")
	}
	sum := sha256.Sum256(raw)
	return hex.EncodeToString(sum[:])[:16], nil
}

// DeriveProjectKEK, per-project KEK'i türetir (§2.3):
// HKDF-SHA-256(ikm=master, salt=hkdfSalt, info=project, L=32).
func DeriveProjectKEK(master []byte, project string) ([]byte, error) {
	if len(master) != 32 {
		return nil, fmt.Errorf("cryptoid.DeriveProjectKEK: master key must be 32 raw bytes")
	}
	kek := make([]byte, 32)
	r := hkdf.New(sha256.New, master, []byte(hkdfSalt), []byte(project))
	if _, err := io.ReadFull(r, kek); err != nil {
		return nil, fmt.Errorf("cryptoid.DeriveProjectKEK: hkdf: %w", err)
	}
	return kek, nil
}

// WrapDEKForKEK, bir DEK'i projenin KEK'i altında WKW1 çerçevesine sarar
// (§2.4). Üretimde bu SUNUCU işidir; Go tarafında yalnızca DR/parite testleri
// fixture üretmek için kullanır. nonce nil → CSPRNG.
func WrapDEKForKEK(master []byte, project string, slot Slot, dek DEK, nonce []byte) ([]byte, error) {
	if err := slot.validate(); err != nil {
		return nil, err
	}
	kek, err := DeriveProjectKEK(master, project)
	if err != nil {
		return nil, err
	}
	if nonce == nil {
		nonce = make([]byte, chacha20poly1305.NonceSizeX)
		if _, rerr := rand.Read(nonce); rerr != nil {
			return nil, fmt.Errorf("cryptoid.WrapDEKForKEK: nonce: %w", rerr)
		}
	}
	if len(nonce) != chacha20poly1305.NonceSizeX {
		return nil, fmt.Errorf("cryptoid.WrapDEKForKEK: nonce must be %d bytes", chacha20poly1305.NonceSizeX)
	}
	aead, err := chacha20poly1305.NewX(kek)
	if err != nil {
		return nil, fmt.Errorf("cryptoid.WrapDEKForKEK: aead: %w", err)
	}
	out := make([]byte, 0, WrapTotalLen)
	out = append(out, WrapMagic...)
	out = append(out, nonce...)
	out = aead.Seal(out, nonce, dek[:], slot.AAD())
	return out, nil
}

// UnwrapDEKWithKEK, bir WKW1 wrap'ini projenin KEK'i altında açar (§2.4).
// AAD = blob AEAD ile AYNI slot bağlaması — bir wrap başka projeye/anahtara/
// versiyona replay edilemez. Her ihlal ErrWrapInvalid (fail-closed, tamper).
func UnwrapDEKWithKEK(master []byte, project string, slot Slot, wrap []byte) (DEK, error) {
	var dek DEK
	if err := slot.validate(); err != nil {
		return dek, err
	}
	if len(wrap) != WrapTotalLen {
		return dek, ErrWrapInvalid
	}
	if string(wrap[:4]) != WrapMagic {
		return dek, ErrWrapInvalid
	}
	kek, err := DeriveProjectKEK(master, project)
	if err != nil {
		return dek, err
	}
	aead, err := chacha20poly1305.NewX(kek)
	if err != nil {
		return dek, fmt.Errorf("cryptoid.UnwrapDEKWithKEK: aead: %w", err)
	}
	nonce := wrap[4 : 4+chacha20poly1305.NonceSizeX]
	ct := wrap[4+chacha20poly1305.NonceSizeX:]
	pt, err := aead.Open(nil, nonce, ct, slot.AAD())
	if err != nil {
		return dek, ErrWrapInvalid
	}
	if len(pt) != dekSize {
		return dek, ErrWrapInvalid
	}
	copy(dek[:], pt)
	return dek, nil
}

File: worker/src/crypto/kek.ts
// MASTER_KEK → per-project KEK → DEK wrap (SPEC §2.2–§2.5). Server-decrypt
// pivotunun kripto çekirdeği:
//   - MASTER_KEK: 64-char hex Worker secret'ı (32 ham bayt). MASTER_KEK_PREV
//     yalnızca rotasyon penceresinde (§2.5) doğrulama/unwrap için.
//   - kid = SHA-256(32 RAW bayt)'ın ilk 16 küçük-harf hex karakteri (§2.2 —
//     hash girdisi ASLA ASCII hex string'i değil, hex-decode edilmiş ham baytlar).
//   - KEK(project) = HKDF-SHA-256(ikm=MASTER_KEK, salt="wapps-secrets/kek/v1",
//     info=project, L=32) (§2.3). İzolat-içi düz Map cache (KV/DO'ya ASLA yazılmaz).
//   - DEK wrap: "WKW1" ‖ nonce(24) ‖ XChaCha20-Poly1305(KEK, nonce, DEK, AAD) —
//     toplam 76 bayt; AAD = blob AEAD ile AYNI slot bağlaması (§2.4).
//
// XChaCha20-Poly1305: WebCrypto'da YOK → @noble/ciphers (pinli, TCB — §2.1 not).

import { xchacha20poly1305 } from "@noble/ciphers/chacha.js";
import { hkdf } from "@noble/hashes/hkdf";
import { sha256 as nobleSha256 } from "@noble/hashes/sha256";
import { bytesToHex, hexToBytes, b64ToBytes, bytesToB64, utf8 } from "./encoding.js";

export const WRAP_RECIPIENT = "worker-kek:v1"; // kapalı küme (§2.4)
export const WRAP_MAGIC = "WKW1";
const HKDF_SALT = "wapps-secrets/kek/v1"; // 20 ASCII bayt, sabit (§2.3)
const NONCE_LEN = 24;
const DEK_LEN = 32;
const TAG_LEN = 16;
export const WRAP_TOTAL_LEN = 4 + NONCE_LEN + DEK_LEN + TAG_LEN; // 76 (§2.4)

/** MasterKey, kurulu bir master anahtar (ham 32 bayt + kid). */
export interface MasterKey {
  kid: string; // 16 küçük-harf hex (§2.2)
  key: Uint8Array; // 32 ham bayt
}

/** KekEnv, master-key secret'larının okunduğu env alt kümesi. */
export interface KekEnv {
  MASTER_KEK?: string;
  MASTER_KEK_PREV?: string;
}

const MASTER_HEX_RE = /^[0-9a-f]{64}$/;

/** kekKid, 32 ham bayttan kid türetir: sha256(raw)[0:16] hex (§2.2). */
export function kekKid(raw: Uint8Array): string {
  return bytesToHex(nobleSha256(raw)).slice(0, 16);
}

/**
 * loadMasterKeys, MASTER_KEK(+PREV)'i env'den yükler. Dönen dizi [current, prev?]
 * sıralıdır; index 0 HER ZAMAN yeni wrap'lerin anahtarıdır. MASTER_KEK eksik/
 * bozuksa null (fail-closed → 503 SERVICE_MISCONFIGURED). Bozuk PREV yok sayılır
 * (aktif anahtar geçerli kaldıkça çalışmaya devam — rotasyon penceresi opsiyonel).
 */
export function loadMasterKeys(env: KekEnv): MasterKey[] | null {
  const cur = (env.MASTER_KEK ?? "").trim().toLowerCase();
  if (!MASTER_HEX_RE.test(cur)) return null;
  const curRaw = hexToBytes(cur);
  const keys: MasterKey[] = [{ kid: kekKid(curRaw), key: curRaw }];
  const prev = (env.MASTER_KEK_PREV ?? "").trim().toLowerCase();
  if (MASTER_HEX_RE.test(prev)) {
    const prevRaw = hexToBytes(prev);
    const prevKid = kekKid(prevRaw);
    if (prevKid !== keys[0].kid) keys.push({ kid: prevKid, key: prevRaw });
  }
  return keys;
}

// İzolat-içi KEK cache (§2.3): düz Map, ASLA KV/DO storage'a yazılmaz. Anahtar
// kid+project — MASTER_KEK rotasyonunda (kid değişir) eski girdiler kullanılmaz olur.
const kekCache = new Map<string, Uint8Array>();

/** deriveProjectKEK, per-project KEK'i türetir (HKDF §2.3; izolat-içi cache). */
export function deriveProjectKEK(master: MasterKey, project: string): Uint8Array {
  const cacheKey = `${master.kid}:${project}`;
  const hit = kekCache.get(cacheKey);
  if (hit) return hit;
  const kek = hkdf(nobleSha256, master.key, utf8(HKDF_SALT), utf8(project), 32);
  kekCache.set(cacheKey, kek);
  return kek;
}

/** __resetKekCache, testler arası izolasyon içindir. */
export function __resetKekCache(): void {
  kekCache.clear();
}

/**
 * slotAAD, hem blob AEAD'inin hem DEK wrap'inin slot bağlaması (§2.1/§2.4):
 * project ‖ 0x00 ‖ keyName ‖ 0x00 ‖ keyVersion(decimal ASCII). Aynı AAD iki
 * katmanda da kullanıldığından bir wrap başka anahtara/projeye/versiyona
 * replay edilemez.
 */
export function slotAAD(project: string, keyName: string, keyVersion: number): Uint8Array {
  const p = utf8(project);
  const k = utf8(keyName);
  const v = utf8(String(keyVersion));
  const out = new Uint8Array(p.length + 1 + k.length + 1 + v.length);
  out.set(p, 0);
  out[p.length] = 0x00;
  out.set(k, p.length + 1);
  out[p.length + 1 + k.length] = 0x00;
  out.set(v, p.length + 1 + k.length + 1);
  return out;
}

/** DekWrap, manifest'te taşınan tek wrap girdisi (§2.4/§2.6). */
export interface DekWrap {
  recipient: string; // "worker-kek:v1"
  kid: string; // 16-hex master-kek id
  wrap: string; // base64(76 bayt)
}

/** WrapError, wrap/unwrap hata sınıfı (§2.4 error contract). */
export class WrapError extends Error {
  constructor(public code: "WRAP_INVALID" | "ALG_UNSUPPORTED", msg?: string) {
    super(msg ?? code);
    this.name = "WrapError";
  }
}

/**
 * wrapDEK, taze DEK'i projenin KEK'i altında sarar (§2.4). nonce parametresi
 * YALNIZCA test determinizmi içindir (frozen vector); üretimde verilmez → CSPRNG.
 */
export function wrapDEK(
  master: MasterKey,
  project: string,
  keyName: string,
  keyVersion: number,
  dek: Uint8Array,
  nonce?: Uint8Array,
): DekWrap {
  if (dek.length !== DEK_LEN) throw new WrapError("WRAP_INVALID", "dek must be 32 bytes");
  const n = nonce ?? crypto.getRandomValues(new Uint8Array(NONCE_LEN));
  if (n.length !== NONCE_LEN) throw new WrapError("WRAP_INVALID", "nonce must be 24 bytes");
  const kek = deriveProjectKEK(master, project);
  const aad = slotAAD(project, keyName, keyVersion);
  const ct = xchacha20poly1305(kek, n, aad).encrypt(dek); // 32+16 bayt
  const out = new Uint8Array(WRAP_TOTAL_LEN);
  out.set(utf8(WRAP_MAGIC), 0);
  out.set(n, 4);
  out.set(ct, 4 + NONCE_LEN);
  return { recipient: WRAP_RECIPIENT, kid: master.kid, wrap: bytesToB64(out) };
}

/**
 * unwrapDEK, bir manifest wrap'ini kid'e uyan master anahtarla (current ya da
 * prev, §2.5 rotasyon penceresi) açar. Bilinmeyen recipient → ALG_UNSUPPORTED;
 * bilinmeyen kid / bozuk çerçeve / AEAD hatası → WRAP_INVALID (503-class,
 * tamper-veya-anahtar-uyuşmazlığı, alert A8 — çağıran fail-closed davranır).
 */
export function unwrapDEK(
  masters: MasterKey[],
  project: string,
  keyName: string,
  keyVersion: number,
  w: DekWrap,
): Uint8Array {
  if (w.recipient !== WRAP_RECIPIENT) throw new WrapError("ALG_UNSUPPORTED", `unknown wrap recipient`);
  const master = masters.find((m) => m.kid === w.kid);
  if (!master) throw new WrapError("WRAP_INVALID", "wrap kid does not match any installed master key");
  let bytes: Uint8Array;
  try {
    bytes = b64ToBytes(w.wrap);
  } catch {
    throw new WrapError("WRAP_INVALID", "wrap not canonical base64");
  }
  if (bytes.length !== WRAP_TOTAL_LEN) throw new WrapError("WRAP_INVALID", "wrap length != 76");
  if (new TextDecoder().decode(bytes.slice(0, 4)) !== WRAP_MAGIC) throw new WrapError("WRAP_INVALID", "bad wrap magic");
  const n = bytes.slice(4, 4 + NONCE_LEN);
  const ct = bytes.slice(4 + NONCE_LEN);
  const kek = deriveProjectKEK(master, project);
  const aad = slotAAD(project, keyName, keyVersion);
  try {
    return xchacha20poly1305(kek, n, aad).decrypt(ct);
  } catch {
    throw new WrapError("WRAP_INVALID", "wrap AEAD open failed");
  }
}
